transformer/external/wasmtransformer.go

```diff
@@ -0,0 +1,242 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package external
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/konveyor/move2kube/common"
+	"github.com/konveyor/move2kube/environment"
+	transformertypes "github.com/konveyor/move2kube/types/transformer"
+	"github.com/second-state/WasmEdge-go/wasmedge"
+	"github.com/sirupsen/logrus"
+	core "k8s.io/kubernetes/pkg/apis/core"
+)
+
+const (
+	wasmEnvDelimiter              = "="
+	detectInputPathWASMEnvKey     = "M2K_DETECT_INPUT_PATH"
+	detectOutputPathWASMEnvKey    = "M2K_DETECT_OUTPUT_PATH"
+	transformInputPathWASMEnvKey  = "M2K_TRANSFORM_INPUT_PATH"
+	transformOutputPathWASMEnvKey = "M2K_TRANSFORM_OUTPUT_PATH"
+)
+
+// WASM implements wasm transformer interface and is used for wasm based transformers
+type WASM struct {
+	Config     transformertypes.Transformer
+	Env        *environment.Environment
+	WASMConfig *WASMYamlConfig
+}
+
+// WASMYamlConfig is the format of wasm transformer yaml config
+type WASMYamlConfig struct {
+	WASMModule string        `yaml:"wasm_module"`
+	EnvList    []core.EnvVar `yaml:"env,omitempty"`
+}
+
+// Init Initializes the transformer
+func (t *WASM) Init(tc transformertypes.Transformer, env *environment.Environment) (err error) {
+	t.Config = tc
+	t.Env = env
+	t.WASMConfig = &WASMYamlConfig{}
+	if err := common.GetObjFromInterface(t.Config.Spec.Config, t.WASMConfig); err != nil {
+		return fmt.Errorf("unable to load config for Transformer %+v into %T . Error: %q", t.Config.Spec.Config, t.WASMConfig, err)
+	}
+
+	return nil
+}
+
+func (t *WASM) prepareEnv() []string {
+	envList := []string{}
+	for _, env := range t.WASMConfig.EnvList {
+		envList = append(envList, env.Name+wasmEnvDelimiter+env.Value)
+	}
+	return envList
+}
+
+// GetConfig returns the transformer config
+func (t *WASM) GetConfig() (transformertypes.Transformer, *environment.Environment) {
+	return t.Config, t.Env
+}
+
+// DirectoryDetect runs detect in each sub directory
+func (t *WASM) DirectoryDetect(dir string) (map[string][]transformertypes.Artifact, error) {
+	vm, err := t.initVm([]string{dir + ":" + dir})
+	if err != nil {
+		return nil, err
+	}
+
+	allocateResult, err := vm.Execute("malloc", int32(len(dir)+1))
+	if err != nil {
+		return nil, err
+	}
+	dirPointer := allocateResult[0].(int32)
+	mod := vm.GetActiveModule()
+	mem := mod.FindMemory("memory")
+	memData, err := mem.GetData(uint(dirPointer), uint(len(dir)+1))
+	if err != nil {
+		return nil, err
+	}
+	copy(memData, dir)
+	memData[len(dir)] = 0
+	directoryDetectOutput, dderr := vm.Execute("directoryDetect", dirPointer)
+	if dderr != nil {
+		err = fmt.Errorf("failed to execute directoryDetect in the wasm module. Error : %s", dderr.Error())
```

Comment:
Use `%w` to wrap the error.
`Error : %w", dderr)`

Reply:
Resolved in d87520b